File: 3341.find-minimum-time-to-reach-last-room-i.cpp
#
# @lc app=leetcode id=3341 lang=cpp
#
# [3341] Find Minimum Time to Reach Last Room I
#

# @lc code=start
class Solution {
public:
    int minTimeToReach(vector<vector<int>>& moveTime) {
        int n = moveTime.size();
        int m = moveTime[0].size();
        vector<vector<long long>> dist(n, vector<long long>(m, LLONG_MAX / 2));
        dist[0][0] = 0;
        using ll = long long;
        priority_queue<tuple<ll, int, int>, vector<tuple<ll, int, int>>, greater<tuple<ll, int, int>>> pq;
        pq.push({0LL, 0, 0});
        int di[4] = {-1, 0, 1, 0};
        int dj[4] = {0, 1, 0, -1};
        while (!pq.empty()) {
            auto [time, i, j] = pq.top(); pq.pop();
            if (time > dist[i][j]) continue;
            for (int d = 0; d < 4; ++d) {
                int ni = i + di[d];
                int nj = j + dj[d];
                if (ni >= 0 && ni < n && nj >= 0 && nj < m) {
                    ll ntime = max(time, (ll)moveTime[ni][nj]) + 1;
                    if (ntime < dist[ni][nj]) {
                        dist[ni][nj] = ntime;
                        pq.push({ntime, ni, nj});
                    }
                }
            }
        }
        return dist[n - 1][m - 1];
    }
};
# @lc code=end

File: 3341.find-minimum-time-to-reach-last-room-i.java
#
# @lc app=leetcode id=3341 lang=java
#
# [3341] Find Minimum Time to Reach Last Room I
#
# @lc code=start
import java.util.*;
class Solution {
    public int minTimeToReach(int[][] moveTime) {
        int n = moveTime.length, m = moveTime[0].length;
        // Validate input
        if (n < 2 || m < 2) return -1;
        for (int[] row : moveTime) {
            if (row.length != m) return -1;
            for (int cell : row) if (cell < 0) return -1;
        }
        int[][] minTime = new int[n][m];
        for (int[] row : minTime) Arrays.fill(row, Integer.MAX_VALUE);
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.offer(new int[]{0, 0, 0}); // {time, x, y}
        minTime[0][0] = 0;
        int[][] dirs = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int time = curr[0], x = curr[1], y = curr[2];
            if (x == n-1 && y == m-1) {
                // Final verification: ensure this is indeed the minimal time
                // as Dijkstra's always expands the node with minimal time
                return time;
            }
            for (int[] d : dirs) {
                int nx = x + d[0], ny = y + d[1];
                if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
                    int arrive = Math.max(time + 1, moveTime[nx][ny]);
                    if (arrive < minTime[nx][ny]) {
                        minTime[nx][ny] = arrive;
                        pq.offer(new int[]{arrive, nx, ny});
                    }
                }
            }
        }
        // Edge case: unreachable
        return -1;
    }
}
# @lc code=end